### PR TITLE
ProxyHandler

### PR DESCRIPTION
**Pull request by Wizzard033**
Welp. This thing took an insane amount of work.

It's done except it only has an integration test and no unit test.
Alfred said he would look into making a unit test.

To run it, just compile and run the server then go to localhost:8080/reverse_proxy
and the integration test uses localhost:8080/simple_proxy

One known bug is with other resources that are linked to from the first proxy'd page
If you click on a link, it will take you to the correct page for sure
However, if you then click another link, suddenly the browser doesn't keep the original URI
So, we made it so that every ProxyHandler is checked during URI resolution
This is only done if the refer header exists (and so we know a user didn't manually type the URL)
It allows the UCLA website to function well, but can give surprisingly bad results in some circumstances
For example, if you have two ProxyHandlers both with bunny.jpg, one website may link to the other's

### CODE REVIEW

**Comment by Wizzard033** [comment]
Found it doesn't compile after I did make clean - fixing now

**Comment by Wizzard033** [comment]
There's still a segfault during "make test"

This is the first commit it started in
https://github.com/UCLA-CS130/TBD/commit/6f7d023f833a3a4c3f88ded40d16c47035935431

**Comment by Wizzard033** [comment]
Fixed - just had to put some extra dependencies on the test

File: Makefile [hunk 1]
```diff
@@ -109,6 +109,9 @@ status_handler_test : status_handler_test.o status_handler.o request_handler.o s
 status_counter_test : status_counter_test.o status_counter.o gmock_main.a
 	$(CXX) $(CPPFLAGS) $(CXXFLAGS) -lpthread $^ -o $@ -lboost_system
 
+redirect_handler_test : redirect_handler_test.o redirect_handler.o request_handler.o gmock_main.a
+	$(CXX) $(CPPFLAGS) $(CXXFLAGS) -lpthread $^ -o $@ -lboost_system
+
 
 TESTS = config_parser_test server_test echo_handler_test \
```
Comment (tincan24) [inline Makefile]: I notice that you guys forgot to add redirect_handler_test to the TESTS variable. Our make test will run all tests included in this variable.

**Comment by jasony0206** [comment]
Integration test looks good, I'll run the unit tests again once Makefile's fixed.

**Comment by jasony0206** [comment]
So we just merged in our multi-threading branch, and that caused some conflicts in server.cc. Have you already submitted your assignment 7? If not, could you let us know when you plan to submit it? I'd imagine the conflict should be minimal. All we changed in server.cc is refactoring `handle_read` in a way that makes it easier to spawn threads.

**Comment by alfredlucero** [comment]
We haven't submitted the assignment 7 yet, but we hope to submit it by tonight after double checking everything and updating READMEs! Yeah we can try and make sure there are no more merge conflicts with server.cc. When do you guys plan to submit as well?

**Comment by tincan24** [comment]
We plan to submit by the end of today as well. Let us know if you guys have any trouble resolving the conflicts.

**Comment by tincan24** [comment]
Our travis failed because it looks like one of the tests in reverse proxy handler test is not passing. Can you guys take a look?

**Comment by Wizzard033** [comment]
I think Jordan and/or Alfred is trying to fix it at the moment

**Comment by alfredlucero** [comment]
Green light all tests passed and hopefully you guys can merge and submit 😄.

**Pull request merged**